### PR TITLE
Snapshots - Support minimal RBAC

### PR DESCRIPTION
**Pull request by salahalsaleh**
Image 2021-02-16 at 6 26 31 PM

### CODE REVIEW

File: cmd/kots/cli/velero.go [hunk 1]
```diff
@@ -30,20 +35,62 @@ func EnsurePermissionsCmd() *cobra.Command {
...
+			if err := snapshot.EnsureVeleroNamespaceConfigMap(veleroNamespace, kotsadmNamespace); err != nil {
 				return err
 			}
 
 			return nil
 		},
 	}
 
-	cmd.Flags().StringP("namespace", "n", "", "namespace in which kots/kotsadm is installed")
+	cmd.Flags().String("kotsadm-namespace", "", "namespace in which kots/kotsadm is installed")
```
Comment (divolgin) [inline cmd/kots/cli/velero.go:60]: why not just `namespace`?

Comment (salahalsaleh) [reply cmd/kots/cli/velero.go:60]: wanted to distinguish between it and `velero-namespace`. no hard opinion here, i can change it back.

File: kotsadm/pkg/snapshot/store.go [hunk 2]
```diff
@@ -549,12 +538,25 @@ func FindBackupStoreLocation() (*velerov1.BackupStorageLocation, error) {
 		return nil, errors.Wrap(err, "failed to get cluster config")
 	}
 
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create clientset")
+	}
+
+	bslNamespace := ""
```
Comment (divolgin) [inline kotsadm/pkg/snapshot/store.go:546]: nitpick: this is called `veleroNamespace` everywhere else

File: pkg/kotsadmsnapshot/backup.go [hunk 3]
```diff
@@ -100,10 +100,6 @@ func CreateApplicationBackup(ctx context.Context, a *apptypes.App, isScheduled b
 	includedNamespaces := []string{appNamespace}
 	includedNamespaces = append(includedNamespaces, kotsKinds.KotsApplication.Spec.AdditionalNamespaces...)
 
-	if os.Getenv("KOTSADM_ENV") == "dev" {
```
Comment (divolgin) [inline pkg/kotsadmsnapshot/backup.go:103]: this intentional?

Comment (salahalsaleh) [reply pkg/kotsadmsnapshot/backup.go:103]: yeah, kotsadm is not included in application backups

File: pkg/kotsadmsnapshot/velero.go [hunk 4]
```diff
@@ -65,18 +110,43 @@ func CheckKotsadmVeleroAccess() (requiresAccess bool, veleroNamespace string, fi
...
+func getMinimalRBACVeleroNamespace(clientset kubernetes.Interface) (string, error) {
+	// a configmap which includes the namespace in which Velero is installed should already be created in minimal rbac mode
+	c, err := clientset.CoreV1().ConfigMaps(os.Getenv("POD_NAMESPACE")).Get(context.TODO(), "kotsadm-velero-namespace", metav1.GetOptions{})
+	if err != nil {
+		if kuberneteserrors.IsNotFound(err) {
+			return "", nil
+		}
+		return "", errors.Wrap(err, "failed to get velero configmap")
+	}
+	return c.Data["veleroNamespace"], nil
```
Comment (divolgin) [inline pkg/kotsadmsnapshot/velero.go:122]: Data can be nil

File: pkg/snapshot/velero.go [hunk 5]
```diff
@@ -1,48 +1,116 @@
...
+
+		_, err := clientset.CoreV1().ConfigMaps(kotsadmNamespace).Create(context.TODO(), newConfigMap, metav1.CreateOptions{})
+		if err != nil {
+			return errors.Wrap(err, "failed to create velero configmap")
+		}
+
+		return nil
+	}
+
+	existingConfigMap.Data["veleroNamespace"] = veleroNamespace
```
Comment (divolgin) [inline pkg/snapshot/velero.go:104]: Data can be nil

File: pkg/kotsadmsnapshot/velero.go [hunk 6]
```diff
@@ -21,41 +23,84 @@ var (
...
+			return
+		}
+		// since this is a minimal rbac installation, kotsadm requires this configmap to know which namespace velero is installed in.
+		// so if it's not found, then the user probably hasn't yet run the command that gives kotsadm access to the namespace velero is installed in,
+		// which will also (re)generate this configmap
+		requiresAccess = true
+		return
+	}
+
+	veleroNamespace := veleroConfigMap.Data["veleroNamespace"]
```
Comment (divolgin) [inline pkg/kotsadmsnapshot/velero.go:65]: Data can be nil

**Review by divolgin — APPROVED** [review]

**Pull request merged**